commercial/app/model/merchandise/books/BookFinder.scala

```diff
@@ -2,67 +2,56 @@ package commercial.model.merchandise.books
 
 import akka.actor.ActorSystem
 import akka.pattern.CircuitBreaker
+import akka.util.Timeout
 import commercial.model.feeds.{FeedParseException, FeedReadException, FeedReader, FeedRequest}
-import common.Logging
+import commercial.model.merchandise.Book
+import common.{AkkaAgent, Logging}
 import conf.Configuration
 import conf.switches.Switches.BookLookupSwitch
-import commercial.model.merchandise.Book
 import play.api.libs.json._
 import play.api.libs.oauth.{ConsumerKey, OAuthCalculator, RequestToken}
 import play.api.libs.ws.{WSClient, WSSignatureCalculator}
-import shade.memcached.{Memcached, MemcachedCodecs, Configuration => MemcachedConfiguration}
 
 import scala.concurrent.duration._
 import scala.concurrent.{ExecutionContext, Future}
-import scala.util.control.NonFatal
 
 class BookFinder(actorSystem: ActorSystem, magentoService: MagentoService) extends Logging {
 
-  private implicit lazy val executionContext = actorSystem.dispatchers.lookup("akka.actor.memcached")
-
-  private lazy val defaultBookDataCache = new MemcachedBookDataCache(actorSystem)
-
-  def findByIsbn(isbn: String,
-                 cache: BookDataCache = defaultBookDataCache,
-                 lookup: String => Future[Option[JsValue]] = magentoService.findByIsbn):
-  Future[Option[Book]] = {
-
-    def cachedBook(bookData: JsValue): Future[Option[Book]] = {
-      Future.successful {
-        bookData match {
-          case JsNull => None
-          case _ =>
-            bookData.validate[Book] fold(
-              invalid => {
-                log.error(Json.stringify(JsError.toJson(invalid)))
-                None
-              },
-              book => Some(book))
-        }
+  private implicit val bookActorExecutionContext: ExecutionContext = actorSystem.dispatchers.lookup("akka.actor.book-lookup")
+  private implicit val bookActorTimeout: Timeout = 0.2.seconds
+  private implicit val magentoServiceImplicit = magentoService
+
+  def findByIsbn(isbn: String): Option[Book] = BookAgent.get(isbn) map { _.as[Book] }
+}
+
+object BookAgent extends Logging {
+
+  private lazy val cache = AkkaAgent(Map.empty[String, JsValue])
+  private lazy val lookupJobs = AkkaAgent(Set.empty[String])
```

Comment:
> Would it make sense to only store pending lookup? (ie: removing it once magento successful request returns).

I was thinking of that but it could lead to race conditions since the `cache` and `lookupJobs` are using different `AkkaAgent`s and there isn't a simple way to ensure updates to one will follow updates to the other made later, we could introduce a third agent but I think we might then suffer from death-by-agent.

>Also how much benefit does adding this extra complexity gives us? How often a request would hit this endpoint while a magento request for the same book is pending?

I don't think it will happen that often? Perhaps 3 times per book? Depends on how much traffic our book pages are generating. Happy to strip it out if it is deemed too much complexity for too little reward - was just trying to be a good citizen 😁